ehrql/debugger.py

```diff
@@ -50,7 +53,39 @@ def show(
     line_no = inspect.getframeinfo(sys._getframe(1))[1]
     label = f" {label}" if label else ""
     print(f"Show line {line_no}:{label}", file=sys.stderr)
-    print(render(element, *other_elements, head=head, tail=tail), file=sys.stderr)
+
+    if DEBUG_QUERY_ENGINE is None:
+        # We're not in debug mode so ignore
+        print(
+            " - show() ignored because we're not running in debug mode", file=sys.stderr
+        )
+        return
+
+    elements = [element, *other_elements]
+
+    is_single_ehrql_object = is_ehrql_object(element) and len(other_elements) == 0
+    is_multiple_series_same_domain = len(elements) > 1 and elements_are_related_series(
+        elements
+    )
+
+    # We throw an error unless show() is called with:
+    # 1. A single ehrql object (series, frame, dataset)
+    # 2. Multiple series drawn from the same domain
+    if is_single_ehrql_object | is_multiple_series_same_domain:
+        print(render(element, *other_elements, head=head, tail=tail), file=sys.stderr)
+    else:
+        raise TypeError(
+            "\n\nshow() can be used in two ways. Either:\n"
+            " - call it with a single argument such as:\n"
+            "     show(dataset)\n"
+            "     show(patients)\n"
+            '     show(clinical_events.date.is_after("2024-01-01"))\n'
+            " - or call it with multiple columns/series:\n"
+            "     show(patients.sex, patients.date_of_birth)\n"
+            '     show(patients.sex == "male", patients.date_of_birth.is_before("2006-01-01"))\n\n'
+            "If passing in multiple arguments, they must all be 'of the same domain', meaning "
+            "you can't combine series from different tables."
```

Comment:
This isn't strictly true - I can combine e.g. `patients.date_of_birth` with `clinical_events.exists_for_patient()` and `ons_deaths.date`; as long as the series are all one row per patient series, they can be drawn from different tables. If they're multiple rows per patient series, they do need to come from the same table.

Reply:
Yes good point - I've updated the error message.